src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -3074,6 +2928,715 @@ void QgsOgrProvider::forceReload()
+      break;
+    default:
+      if ( mLayerIndex >= 0 )
+      {
+        if ( mDriverName != "SQLite" )
+        {
+          if ( ( !mLayerName.contains( "(" ) ) && ( !mLayerName.endsWith( QString( ")" ) ) ) )
+          {
+            if ( ( mGeometryName.isNull() ) || ( mGeometryName.isEmpty() ) )
+            {
+              mOgrGetType = 1; // read by layer-index [KML, GML with one geometry]
+            }
+            else
+            {
+              mOgrGetType = 2; // read by layer and geomety-name [GML with more than one geometry]
+            }
+          }
+          else
+          {
+            if ( ( mLayerName.contains( "(" ) ) && ( mLayerName.endsWith( QString( ")" ) ) ) )
+            {
+              mOgrGetType = 2; // read by layer and geomety-name [GML with more than one geometry]
+            }
+            else
+            {
+              if ( ( mGeometryName.isNull() ) || ( mGeometryName.isEmpty() ) )
+              {
+                mOgrGetType = 1; // read by layer-index [KML, GML with one geometry]
+              }
+              else
+              {
+                mOgrGetType = 2; // read by layer and geomety-name [GML with more than one geometry]
+              }
+            }
+          }
+        }
+        else
+        {
+          // SQLite.  'table_name(geometry_column_name)'
+          mOgrGetType = 0; // read by layer-name
+        }
+      }
+      mSubLayerString = QString( "%1:%2:%3:%4:%5:%6:%7" ).arg( mLayerIndex ).arg( mLayerName ).arg( mFeaturesCounted ).arg( QgsOgrProviderUtils::wkbGeometryTypeName( mOgrGeometryTypeFilter ) ).arg( sGeometryName ).arg( mGeometryIndex ).arg( mOgrGetType );
+      break;
+  }
+  if ( ( ogrDataSource ) && ( !mSubLayerString.isNull() ) )
+  {
+    ogrLayer = QgsOgrProviderUtils::OGRGetSubLayerStringWrapper( ogrDataSource, mSubLayerString );
+    if ( ogrLayer )
+    {
+      return ogrLayer;
+    }
+    mSubLayerString = QString::null;
+    mOgrGetType = -1;
+  }
+  if ( ( ogrDataSource ) && ( mSubLayerString.isNull() ) )
+  {
+    // Retrieving the Layer has failed, attempting again to read the DataSource.
+    long layer_count = 0;
+    int i_count_id = 0;
+    bool ok;
+    QString s_sublayer = QString::null;
+    if ( !mValid )
+    {
+      // this is being opened the first time as a check, override the valid check
+      layer_count = QgsOgrProviderUtils::OGRGetLayerCountWrapper( ogrDataSource );
+    }
+    // Problem: since gdal 2.*, GetLayer with an index will not work as before 2.*
+    // - a table with multiple geometries is considered 1 layer (before each geometry was 1 layer)
+    // - GetLayer with Name must be used to work correctly, assuming the name is unique
+    // --> which for GML, KML may not be the case
```

Comment:
OK, I have removed GML, although if a GNL file is created manually
* it is possible, this the 'may'